walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,118 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    /* Initializing an array that will provide information for each minute in a day. 
+       dayMinutes[j] will represent what is happening in minute j, where j is
+       from 0 to 3599. */
+    ArrayList<MinuteDetails> dayMinutes = new ArrayList<>();
+    for (int i = 0; i <= TimeRange.END_OF_DAY  + 1; i++) {
+      /* Initially, consider that each minute of the day is available for a metting. 
+         If that is so, the next available minute would be the next one. */ 
+      dayMinutes.add(new MinuteDetails(true, i + 1));
+    }
+    /* Set in the array the minutes in which all mandatory attendees are not available. */
+    setMandatoryAttendeesAvailability(dayMinutes, events, request.getAttendees());
+
+    return findAavailableTimeRanges(dayMinutes, (int) request.getDuration());
+  }
+
+  /* Set in dayMinutes the minutes in which the meeting can't be held and the next availability
+     field for every minute. */
+  private void setMandatoryAttendeesAvailability(ArrayList<MinuteDetails> dayMinutes,
+      Collection<Event> events, Collection<String> attendees) {
+        for (Event event : events) {
+          for (String person : attendees) {
+            if (event.attends(person)) {
+              /* Set all the minutes of the current event as unavailabe in dayMinutes. */
+              int start = event.getWhen().start();
+              int end = event.getWhen().end();
+              /* Update the next availability for the minute before the event. */
+              if (start > 0 && dayMinutes.get(start - 1).getNextAvailableMinute() < end) {
+                dayMinutes.get(start - 1).setNextAvailableMinute(end);
+              }
+              for (int i = start; i < end; i++) {
+                if (i <= TimeRange.END_OF_DAY) {
+                  dayMinutes.get(i).setMandatoryAvailability(false);
+                  /* Update the next availability for the current minute only if the setted
+                  availability does not point to an even later time. */
+                  if (dayMinutes.get(i).getNextAvailableMinute() < end) {
+                    dayMinutes.get(i).setNextAvailableMinute(end);    
+                  }
+                } else {
```

Comment:
Optional: It's often a good idea to reduce levels of indentation by adopting the `early return` pattern:
```
for (x in list) {
  if (x is bad) {
    break;
  }
  // code here assumes x is good.
}
```

You can apply this pattern twice in this function to get rid of 2 levels of indentation.

Note this is not in the style guide afaik, though I believe most developers find early return more readable (and is de facto enforced internally in most of Google's internal codebase).

Reply:
Will do